### PR TITLE
New feature: Search screen

### PR DESCRIPTION
**Pull request by nqhhdev**

### CODE REVIEW

File: lib/config/go_routes/go_router.dart [hunk 1]
```diff
@@ -97,213 +98,197 @@ class TwakeRoutes {
...
                 ],
               ),
             ],
           ),
         ],
       ),
+    ],
+  );
+
+  StatefulShellBranch get _searchShellBranch => StatefulShellBranch(
```
Comment (hoangdat) [inline lib/config/go_routes/go_router.dart:246]: why not see `empty` and `_roomIdRoute` in search branch?

File: lib/presentation/extensions/send_image_extension.dart [hunk 2]
```diff
@@ -9,13 +9,13 @@ import 'package:fluffychat/presentation/extensions/asset_entity_extension.dart';
...
 import 'package:path_provider/path_provider.dart';
-import 'package:photo_manager/photo_manager.dart';
 import 'package:matrix/src/utils/file_send_request_credentials.dart';
+import 'package:photo_manager/photo_manager.dart';
 
 typedef TransactionId = String;
 
 typedef FakeImageEvent = SyncUpdate;
-extension SendImage on Room {
+extension SendImageExtension on Room {
```
Comment (hoangdat) [inline lib/presentation/extensions/send_image_extension.dart:18]:
```suggestion
extension SendFileExtension on Room {
```

Comment (hoangdat) [reply lib/presentation/extensions/send_image_extension.dart:18]: `Future<String?> sendImageFileEvent( ` -> to `Future<String?> sendFileEvent(`

File: lib/config/go_routes/go_router.dart [hunk 3]
```diff
@@ -96,205 +98,200 @@ class TwakeRoutes {
...
-                    pageBuilder: (context, state) => NoTransitionPage(
-                      child: AdaptiveScaffoldRoute(
-                        body: const ContactsTab(),
-                        secondaryBody: Chat(
-                          key: ValueKey(state.pathParameters['roomid'] ?? ""),
-                        ),
-                      ),
+                ),
+                GoRoute(
+                  path: 'emptyChat',
```
Comment (hoangdat) [inline lib/config/go_routes/go_router.dart:158]: Change it to `draftChat`

File: lib/config/go_routes/go_router.dart [hunk 4]
```diff
@@ -96,205 +98,200 @@ class TwakeRoutes {
...
+                  pageBuilder: (context, state) => NoTransitionPage(
+                    child: AdaptiveScaffoldRoute(
+                      body: const ChatList(),
+                      secondaryBody: EmptyChat(state: state),
```
Comment (hoangdat) [inline lib/config/go_routes/go_router.dart:162]: `DraftChat`

File: lib/config/go_routes/go_router.dart [hunk 5]
```diff
@@ -38,14 +39,15 @@ import 'package:matrix/matrix.dart';
...
-  final _navChatNavigatorKey = GlobalKey<NavigatorState>(debugLabel: 'Chat');
+  final _navRoomsNavigatorKey = GlobalKey<NavigatorState>(debugLabel: 'rooms');
   final _navContactsNavigatorKey = GlobalKey<NavigatorState>(debugLabel: 'Contacts');
   final _navStoriesNavigatorKey = GlobalKey<NavigatorState>(debugLabel: 'Stories');
 
   static const List<String> shellBranch = [
     '/rooms',
     '/contacts',
     '/stories',
+    '/search',
```
Comment (hoangdat) [inline lib/config/go_routes/go_router.dart:50]: Really?

Comment (nqhhdev) [reply lib/config/go_routes/go_router.dart:50]: I removed it

File: lib/config/go_routes/go_router.dart [hunk 6]
```diff
@@ -329,7 +324,9 @@ class TwakeRoutes {
   List<RouteBase> get _settingsRoutes => [
     GoRoute(
       path: 'notifications',
-      builder: (context, state) => const SettingsNotifications(),
+      pageBuilder: (context, state) => const NoTransitionPage(
+        child: SettingsNotifications()
+      ),
```
Comment (hoangdat) [inline lib/config/go_routes/go_router.dart:329]: can you express sth about it?

File: lib/pages/chat_list/chat_list.dart [hunk 7]
```diff
@@ -132,6 +133,7 @@ class ChatListController extends State<ChatList>
     .client
     .rooms
     .where(getRoomFilterByActiveFilter(_activeFilterAllChats))
+    .where((room) => room.isShowInChatList())
```
Comment (hoangdat) [inline lib/pages/chat_list/chat_list.dart:136]: IMO, we dont need it now.

File: lib/presentation/extensions/send_file_extension.dart [hunk 8]
```diff
@@ -220,4 +220,24 @@ extension SendImage on Room {
...
+  bool _isDirectChatHaveMessage() {
+    return isDirectChat && _isLastEventInRoomIsMessage();
+  }
+
+  bool _isLastEventInRoomIsMessage() {
+    return [
+      EventTypes.Message,
+      EventTypes.Sticker,
+      EventTypes.Encrypted,
+    ].contains(lastEvent?.type);
```
Comment (hoangdat) [inline lib/presentation/extensions/send_file_extension.dart:241]: remove it

File: lib/presentation/extensions/send_file_extension.dart [hunk 9]
```diff
@@ -220,4 +220,24 @@ extension SendImage on Room {
   User? getUser(mxId) {
     return getParticipants().firstWhereOrNull((user) => user.id == mxId);
   }
+
+  bool isShowInChatList() {
```
Comment (hoangdat) [inline lib/presentation/extensions/send_file_extension.dart:224]: we need it more @sherlockvn?

File: lib/presentation/extensions/send_image_extension.dart (hunk 2 above)
Comment (hoangdat) [reply lib/presentation/extensions/send_image_extension.dart]: you forgot it, change the method name also

File: lib/config/go_routes/go_router.dart (hunk 6 above)
Comment (nqhhdev) [reply lib/config/go_routes/go_router.dart]: Just testing, I reverted the old code

**Review by hoangdat — APPROVED** [review]

**Review by sherlockvn — APPROVED** [review]

File: lib/presentation/extensions/send_file_extension.dart (hunk 9 above)
Comment (sherlockvn) [reply lib/presentation/extensions/send_file_extension.dart]: no need

**Review by drminh2807 — APPROVED** [review]

**Pull request merged**